### PR TITLE
Automate creating github release

### PR DESCRIPTION
**Pull request by DennisDenuto**
- Uses goreleaser to upload and generate draft release notes with assets attached, and preliminary release notes generated

Authored-by: Dennis Leon <leonde@vmware.com>

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    tags:
+      - 'v*'
+
+jobs:
+  goreleaser:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      -
```
Comment (pivotaljohn) [inline .github/workflows/release.yml]: Just a question. This is legit YAML, but it's less common to put a newline between the "array item" and the contained "map".

I'm down for trying something different to see how it looks/feels.

At this moment, I'm finding it _less_ readable as the dashes (`-`)  seem slightly detached from the content of their step, by my eye... making it distracting for me.

It's one of those strange nits... has zero affect on the correctness, but some kind of meaningful impact to _this_ human.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        if: startsWith(github.ref, 'refs/tags/')
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          result-encoding: string
+          script: |
+            var crypto = require('crypto');
+            const { owner, repo } = context.repo;
+
+            // https://docs.github.com/en/rest/reference/repos#list-releases
+            // https://octokit.github.io/rest.js/v18#repos-list-releases
```
Comment (pivotaljohn) [inline .github/workflows/release.yml:46]: 🙏🏻  — really appreciate it when folks point to the API docs.

File: .goreleaser.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# This is an example .goreleaser.yml file with some sane defaults.
```
Comment (pivotaljohn) [inline .goreleaser.yml:1]: Thoughts on pruning configuration that (re)sets to defaults?

File: .goreleaser.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  # Default is empty
+  sort: asc
+
+  filters:
+    # Commit messages matching the regexp listed here will be removed from
+    # the changelog
+    # Default is empty
+    exclude:
+      - '^docs:'
+      - typo
```
Comment (pivotaljohn) [inline .goreleaser.yml]: nit: end-of-file newline?

File: examples/version-constraint/config.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#@ load("@ytt:version", "version")
+#@ load("@ytt:overlay", "overlay")
```
Comment (pivotaljohn) [inline examples/version-constraint/config.yml]: superfluous. 
```suggestion
```

File: examples/version-constraint/config.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+
+#@ version.require_at_least("0.0.0")
```
Comment (pivotaljohn) [inline examples/version-constraint/config.yml]: nit: end-of-line newline?

File: pkg/version/version.go [hunk 7]
```diff
@@ -3,6 +3,4 @@
 
 package version
 
-const (
-	Version = "0.34.0"
-)
+var Version = "develop"
```
Comment (pivotaljohn) [inline pkg/version/version.go:6]: Question to understand: what does setting this to non-semver do for us?

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        with:
+          fetch-depth: 0
+      -
+        name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.16.0
+      -
+        name: Run GoReleaser
+        uses: goreleaser/goreleaser-action@5e15885530fb01d81d1f24e8a6f54ebbd0fed7eb
```
Comment (pivotaljohn) [inline .github/workflows/release.yml:27]: yes: immutable SHA. 👍🏻 

Perhaps mention intended version?

Also, 2.6.1 is out; should we grab that upgrade right now? 🤷🏻

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        if: startsWith(github.ref, 'refs/tags/')
+        with:
+          version: 0.162.0
+          args: release --rm-dist --debug
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (pivotaljohn) [inline .github/workflows/release.yml:33]: What are the permissions/scopes of this token?

**Review by pivotaljohn — COMMENTED** [review]
In general, looks good.

Left some questions to speed a synchronous review.

Thanks for putting this together!!!

File: .github/workflows/release.yml (hunk 1 above)
Comment (DennisDenuto) [reply .github/workflows/release.yml]: Yeah thats fair. I've changed it to not have a new line after the array item.

File: .github/workflows/release.yml (hunk 8 above)
Comment (DennisDenuto) [reply .github/workflows/release.yml:27]: I've added a comment to specify which version it belongs to. I haven't bumped it yet. I would like to test that version bump first although reading the release notes it should be really safe (they seem like chore related changes). Assuming everything still works, i'll bump all the tools to keep goreleaser action in sync across the tools.

I'd like to submit that bump as a separate PR. I plan on working on it next dev day.

File: .github/workflows/release.yml (hunk 9 above)
Comment (DennisDenuto) [reply .github/workflows/release.yml:33]: They are above on line 12. The token permissions are set to allow 'write' on 'content' resource. I added a comment to refer to the docs around setting permissions.

File: .github/workflows/release.yml (hunk 2 above)
Comment (DennisDenuto) [reply .github/workflows/release.yml:46]: 🙇

File: .goreleaser.yml (hunk 3 above)
Comment (DennisDenuto) [reply .goreleaser.yml:1]: I have found it re-assuring / valuable when reading this file, to see the different configuration options set (even if they are set to their default).

i.e. seeing `draft: true` in this file. 
If I were to remove this line, I think I would check the docs every time to make sure that the default is to publish a draft.

File: pkg/version/version.go (hunk 7 above)
Comment (DennisDenuto) [reply pkg/version/version.go:6]: This was primarily to make it consistent with feedback received for imgpkg.

I am not sure it does technically much for us. It acts more as a placeholder that the version comes from 'somewhere else'. And that somewhere else is our build tooling (via go's LDD build flags).

**Comment by DennisDenuto** [comment]
Thanks for the feedback. I've addressed most of it (i think)

Any feedback given here, I've also ported over to kapp and kbld. i'll eventually get it over to imgpkg too.

File: examples/version-constraint/config.yml (hunk 5 above)
Comment (DennisDenuto) [reply examples/version-constraint/config.yml]: done

**Pull request merged**